libsolidity/ast/TypeProvider.h

```diff
@@ -0,0 +1,213 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/Types.h>
+
+#include <array>
+#include <map>
+#include <memory>
+#include <utility>
+
+namespace dev
+{
+namespace solidity
+{
+
+/**
+ * API for accessing the Solidity Type System.
+ *
+ * This is the Solidity Compiler's type provider. Use it to request for types. The caller does
+ * <b>not</b> own the types.
+ *
+ * It is not recommended to explicitly instantiate types unless you really know what and why
+ * you are doing it.
+ */
+class TypeProvider
+{
+public:
+	TypeProvider();
+	TypeProvider(TypeProvider&&) = default;
+	TypeProvider(TypeProvider const&) = delete;
+	TypeProvider& operator=(TypeProvider&&) = default;
+	TypeProvider& operator=(TypeProvider const&) = delete;
+	~TypeProvider() = default;
+
+	/**
+	 * Resets state of this TypeProvider to initial state, wiping all mutable types.
+	 *
+	 * This invalidates all dangling pointers to types provided by this TypeProvider.
+	 */
+	void reset();
+
+	/**
+	 * Global TypeProvider instance.
+	 */
+	static TypeProvider& get()
```

Comment:
I wonder whether it'd be nicer to make this one private and all the ones below static as well and let them use this one - e.g.
``static BoolType const* boolType() const noexcept { return &get().m_boolType; }``
and then
``auto boolType = TypeProvider::boolType();`` instead of ``auto boolType = TypeProvider::get().boolType();``

Reply:
I agree - if we stay with singleton, then it would be better to avoid the `get` call.